### PR TITLE
[Feat][MoE] Add Qwen3MoEOp and refactor permute/unpermute to padded layout.

### PR DESCRIPTION
**Pull request by michaelwithu**
## Summary

This PR implements `Qwen3MoEOp`, a full MoE FFN operator for Qwen3/Qwen2/Mixtral-style
models, and refactors `MoePermuteOp`/`MoeUnpermuteOp` to a padded layout that eliminates
all CPU–GPU syncs in the forward path.

### New op: `Qwen3MoEOp`

Chains the following TileOPs kernels into a single forward pass with zero CPU loops and
zero `.cpu()` calls:

FusedTopKOp → MoePermuteOp → GroupedGemmOp (gate+up) → SiluAndMulOp → GroupedGemmOp (down) → MoeUnpermuteOp

Supports `softmax` (Qwen3/Qwen2/Mixtral) and `sigmoid` (DeepSeek-V3) routing, with
optional renormalization.

### Padded layout refactor: `MoePermuteOp` / `MoeUnpermuteOp`

The NT grouped-GEMM kernel requires each expert's token block to start on a `block_m=64`
boundary. Previously this padding was computed on CPU in a Python for-loop with a
`.cpu()` sync. It is now computed entirely on GPU inside the scan kernel:

- `MoePermuteOp` outputs `perm_h_pad [padded_batch_sum, H]` + `fwd_idx [T×K]`
  (flat_idx → padded_slot) instead of a non-padded buffer + inverse index
- `MoeUnpermuteOp` accepts `mm2_pad [padded_batch_sum, H]` + `fwd_idx`, directly
  indexing the padded GEMM output without an intermediate scatter step
- The forward mapping `fwd_idx` replaces the old inverse mapping, which was incorrect
  for large E (failed when the permutation was not a self-inverse)

### EP scope

The current implementation covers the **local-expert compute** portion of Expert
Parallelism: `Qwen3MoEOp` processes only the experts whose weights are present on the
current device. This is the exact kernel that runs on each rank inside an EP All-to-All
loop.

Multi-GPU EP dispatch (All-to-All communication + `expert_map`-based token filtering)
is **not yet implemented**. The `expert_map` parameter is accepted and stored for future
use. As a result:
- **Supported now**: single-GPU models where all expert weights fit in HBM
  (e.g. Qwen3-30B-A3B ~64 GB, Mixtral-8x7B)
- **Not supported yet**: models requiring EP (e.g. Qwen3-235B-A22B ~700 GB,
  DeepSeek-V3 ~2 TB)

### E=128, K=8, H=2048, F=1024 (softmax, no renorm — Qwen3-MoE style)

| T | TileOPs (ms) | TileOPs (TFLOPS) | vLLM (ms) | vLLM (TFLOPS) | PyTorch (ms) | vs vLLM | vs PyTorch |
|---|---|---|---|---|---|---|---|
| 512  | 0.61 | 83.9  | 0.59 | 87.1  | 12.49 | -3%   | 20×   |
| 2048 | 1.01 | 205.1 | 1.28 | 161.4 | 17.93 | +21%  | 18×   |
| 4096 | 1.58 | 261.7 | 1.29 | 318.8 | 24.03 | -18%  | 15×   |

### E=256, K=8, H=2048, F=1024 (softmax, renorm — Qwen3.5-MoE style)

| T | TileOPs (ms) | TileOPs (TFLOPS) | vLLM (ms) | vLLM (TFLOPS) | PyTorch (ms) | vs vLLM | vs PyTorch |
|---|---|---|---|---|---|---|---|
| 512  | 1.20 | 42.9  | 0.83 | 62.4  | 25.09 | -31%  | 21×   |
| 2048 | 1.50 | 137.6 | 1.03 | 200.0 | 29.59 | -31%  | 20×   |
| 4096 | 2.10 | 196.4 | 1.43 | 289.3 | 36.45 | -32%  | 17×   |

### E=256, K=8, H=2048, F=1024 (sigmoid, renorm — DeepSeek-V3 style)

| T | TileOPs (ms) | TileOPs (TFLOPS) | PyTorch (ms) | vs PyTorch |
|---|---|---|---|---|
| 512  | 1.20 | 42.8  | 25.12 | 21×  |
| 2048 | 1.50 | 137.2 | 29.91 | 20×  |
| 4096 | 2.10 | 196.7 | 36.50 | 17×  |

> vLLM baseline not available for sigmoid routing.

## Performance Gap Analysis

Profiling shows `GroupedGemmOp` dominates end-to-end latency (79–84%):

| T | E | gemm_gu | gemm_dn | GEMM total | e2e | GEMM % |
|---|---|---|---|---|---|---|
| 512  | 128 | 0.78ms | 0.42ms | 1.20ms | 1.42ms | 84% |
| 4096 | 128 | 1.77ms | 0.99ms | 2.76ms | 3.46ms | 79% |
| 4096 | 256 | 2.48ms | 1.41ms | 3.89ms | 4.68ms | 83% |

The gap vs vLLM at large T/E is caused by two factors:
1. **vLLM uses a pre-tuned H200 Triton config** (`E=128,N=1024,device_name=NVIDIA_H200.json`),
   while TileOPs uses a fixed config (`block_n=256, block_k=64, num_stages=2`)
2. **Padding overhead**: the padded layout adds up to `E × block_m` extra rows that
   still participate in GEMM computation

## Next Steps

- [ ] **GroupedGemmOp autotune for H200/H100**: tune `block_m/n/k`, `num_stages` for
  MoE-scale batch sizes
- [ ] **Reduce padding overhead**: skip padding rows in the NT kernel via mask or
  adjusted launch config
- [ ] **Multi-GPU EP dispatch**: implement All-to-All + `expert_map` token filtering in
  `Qwen3MoEOp.forward` to support 235B+ models
- [ ] **TP support**: split F dimension across ranks with AllReduce after down projection

## Test Plan

- [x] 25 unit tests passing: `test_moe_permute` (9), `test_moe_unpermute` (9),
  `test_moe_qwen3_moe` (7, including smoke + full E=128/256 configs)
- [x] Correctness validated against PyTorch per-expert reference (rtol=1e-2, atol=1e-2)
- [x] Correctness cross-checked against vLLM `fused_experts` for softmax routing

### CODE REVIEW

File: tileops/ops/moe/qwen3_moe.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+from tileops.ops.moe.permute import MoePermuteOp
+from tileops.ops.moe.unpermute import MoeUnpermuteOp
+
+from ..op import Op
+
+__all__ = ["Qwen3MoEOp"]
+
+# NT kernel default block_m (grouped_gemm.py: (False, True): {"block_m": 64, ...}).
+# Each expert's row slice must start on this boundary for correct tile→expert mapping.
+_BLOCK_M_NT = 64
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tileops/ops/moe/qwen3_moe.py]: Good catch. Adopted, but used a lighter approach: instead of instantiating a dummy `GroupedGemmOp` (which triggers JIT compilation), we import `_DEFAULT_CONFIGS` directly from the kernel module and read `block_m` at module load time — zero overhead.

```python
from tileops.kernels.grouped_gemm.grouped_gemm import _DEFAULT_CONFIGS as _GEMM_DEFAULT_CONFIGS
_BLOCK_M_NT: int = _GEMM_DEFAULT_CONFIGS[(False, True)]["block_m"]
```

This way `_BLOCK_M_NT` stays in sync with the kernel default automatically. Fixed in 59b1919.

**Review by superAngGao — COMMENTED** [review]
## Code Review

### Critical: Bug Fix Deserves More Visibility

The PR description mentions that `fwd_idx` "replaces the old inverse mapping, which was incorrect for large E." This is actually a **serious correctness fix** and deserves more prominent treatment.

In the old scan kernel, `inv_permuted_idx[slot] = idx` writes using `slot` as the index, but the unpermute kernel reads with `inv_permuted_idx[flat_idx]` using `flat_idx` as the index. These are different index spaces — the code only worked by coincidence when the permutation happened to be approximately self-inverse. The new `fwd_idx[idx] = padded_slot` correctly uses the same index space (flat_idx) for both write and read.

Recommend explicitly calling this out as a bug fix in the PR description / commit message, not just a refactor.

### Design Concerns

**1. `expert_map` — dead code placeholder**

`Qwen3MoEOp.__init__` accepts `expert_map` and stores it as `self.expert_map`, but it is never used. If a caller passes a non-None value, they will silently get incorrect behavior (thinking EP is active when it is not). Consider raising `NotImplementedError` when a non-None value is passed.

**2. `padded_batch_sum` default in `MoeUnpermuteOp`**

`unpermute.py` defaults `padded_batch_sum` to `num_tokens * top_k`. If a user passes a padded buffer from `MoePermuteOp` but forgets to specify this parameter, the kernel will be compiled with the wrong tensor shape. Consider removing the default and making it a required parameter to force explicit intent.

**3. `Qwen3MoEOp` naming**

The op supports both softmax (Qwen3/Qwen2/Mixtral) and sigmoid (DeepSeek-V3) routing. The name `Qwen3MoEOp` is model-specific and may confuse users who want to use it for Mixtral or DeepSeek.

**4. `kernel_map` parameter silently ignored**

`Qwen3MoEOp.default_kernel_map` returns `{}` and `dispatch_kernel` is never called. The `kernel_map` constructor parameter is accepted but has no effect.

**5. `torch.library.custom_op` not registered**

All MoE kernels and ops in this PR — `Qwen3MoEOp` (new), as well as existing `MoePermuteOp`, `MoeUnpermuteOp`, `FusedTopKOp`, and `GroupedGemmOp` — are not registered via `torch.library.custom_op` + `register_fake`. Within `Qwen3MoEOp.forward`, the sub-op `SiluAndMulOp` IS registered while the other 5 sub-ops are not, resulting in inconsistent dispatch paths.
